### ISSUE TITLE
Accuracy calculation

### ISSUE THREAD

**Issue by Anabik** [KevinMusgrave/pytorch-metric-learning#334]
I am learning how to use this toolkit. My goal is to develop a deep model which will provide linearly separable feature for  dealing with three class classification problem. I divide my dataset into train, valid and test set and all set contains data from three classes. I want to train a KNN with the embedding received from the train data and see the classification performance on test data.

My model has three parts trunk, embedder and classifier. The embedder is a single linear layer based on which NPairsLoss is computed and classifier is placed on top of embedder. I am using CrossEntropyLoss() for classifier, set sampler as samplers.MPerClassSampler(train_dataset.targets, m=2,batch_size = 6, length_before_new_iter=len(train_dataset)). I am training the model with trainer.train() and receive  accuracy around 90%. The following text is copied from the output screen:
INFO:PML:New best accuracy! 0.9065040650406504

However, when I load (with trunk.load_state_dict(),  and embedder.load_state_dict() ) the best saved models and compute the image embeddings (with different python program after models) and run KNN the accuracy is very poor (around 0.33). Can you explain the reason? Specifically, I want to know the computation behind computing accuracy? 

In case, if you need to more please reply me.

Thanks

**Comment by KevinMusgrave** [comment]
If you're using the default Tester and AccuracyCalculator settings, then it does the following:

## Normalizes all embeddings 

In your python program, make sure to do 

```embeddings = torch.nn.functional.normalize(embeddings)``` before computing k-nn. 

Or if you don't want to normalize embeddings, then set ```normalize_embeddings = False```:

```python
tester = testers.GlobalEmbeddingSpaceTester(<your_other_args>, normalize_embeddings=False)
```


## Computes k-nn for the validation set using the validation set as the search space. 

Your python program might be using both the training set and validation set as the search space. If you want to do this with PML then set ```splits_to_eval```:

```python
# use both the train and val set as the search space when computing k-nn accuracy for the val set
splits_to_eval = [("val", ["train", "val"])]
end_of_epoch_hook = hooks.end_of_epoch_hook(<your_other_args>, splits_to_eval=splits_to_eval)
```


## Uses ```mean_average_precision_at_r``` as the default accuracy metric

This is different from 1-nearest-neighbor accuracy. If you want to use nearest neighbor accuracy, you can set ```primary_metric``` to ```precision_at_1```:

```python
hooks = logging_presets.get_hook_container(record_keeper, primary_metric="precision_at_1")
```

**Comment by Anabik** [comment]
Thank you Kevin Musgrave.  I can understand.
